adaptive_sampler.go

```diff
@@ -0,0 +1,160 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jaeger
+
+import (
+	"fmt"
+	"sync"
+	"time"
+	"github.com/uber/jaeger-client-go/utils"
+	"math"
+)
+
+const (
+	defaultProbability = 0.001
+)
+
+// struct for debugging purposes only, keeps track of throughput and number of samples
+type metadata struct {
+	throughput int64
+	samples    int64
+}
+
+type AdaptiveSampler struct {
+	samplers            map[string]*ProbabilisticSampler
+	upperbound          map[string]utils.RateLimiter
+	targetQPS           float64
+	probabilities       map[string]float64
+	samples             map[string]int64
+	lastTick            time.Time
+	mux                 sync.Mutex
+	calculationInterval time.Duration
+
+	totalSampled map[string]*metadata
+	start        time.Time
+}
+
+func NewAdaptiveSamplerV2(targetQPS float64) (*AdaptiveSampler, error) {
+	sampler := &AdaptiveSampler{
+		samples:             make(map[string]int64),
+		samplers:            make(map[string]*ProbabilisticSampler),
+		upperbound:          make(map[string]utils.RateLimiter),
+		probabilities:       make(map[string]float64),
+		targetQPS:           targetQPS,
+		calculationInterval: time.Duration(1000000.0/targetQPS) * time.Microsecond * 2, // Nyquist rate
+		start:               time.Now(),
+		totalSampled:        make(map[string]*metadata),
+	}
+	go sampler.run()
+	go sampler.log()
+	return sampler, nil
+}
+
+func (a *AdaptiveSampler) run() {
+	a.lastTick = time.Now()
+	ticker := time.NewTicker(a.calculationInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			a.calculate()
+		//case <-t.close:
+		//	return
+		}
+	}
+}
+
+func (a *AdaptiveSampler) log() {
+	ticker := time.NewTicker(time.Second * 60)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			a.mux.Lock()
+			print(fmt.Sprintf("duration: %v\n", time.Now().Sub(a.start)))
+			for operation, d := range a.totalSampled {
+				print(fmt.Sprintf("operation: %s throughput: %v samples: %v\n", operation, d.throughput, d.samples))
+			}
+			a.mux.Unlock()
+		}
+	}
+}
+
+func (a *AdaptiveSampler) IsSampled(id TraceID, operation string) (bool, []Tag) {
+	a.mux.Lock()
+	defer a.mux.Unlock()
+	if _, ok := a.samples[operation]; !ok {
+		a.samples[operation] = 0
+	}
+	if _, ok := a.totalSampled[operation]; !ok {
+		a.totalSampled[operation] = &metadata{}
+	}
+	var sampler *ProbabilisticSampler
+	var ok bool
+	if sampler, ok = a.samplers[operation]; !ok {
+		sampler, _ = NewProbabilisticSampler(defaultProbability)
+		a.probabilities[operation] = defaultProbability
+		a.samplers[operation] = sampler
+
+		upperbound := utils.NewRateLimiter(a.targetQPS, math.Max(a.targetQPS, 1.0))
+		a.upperbound[operation] = upperbound
+	}
+	sampled, tags := sampler.IsSampled(id, operation)
+	a.totalSampled[operation].throughput += 1
```

Comment:
should replace a.totalSampled[operation].throughput += 1 with a.totalSampled[operation].throughput++